pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,493 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitortool
+
+import (
+	"net"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+	corev1 "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	config "antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha1"
+)
+
+var (
+	icmpSeq uint32
+)
+
+const (
+	// For non-privileged
+	IPv4ProtocolICMP = "udp4"
+	IPv6ProtocolICMP = "udp6"
+	// For privileged
+	IPv4ProtocolICMPRaw = "ip4:icmp"
+	IPv6ProtocolICMPRaw = "ip6:ipv6-icmp"
+	// IP Protocol
+	IPProtocol = "ip"
+	// ProtocolICMP is the ICMP protocol number.
+	ProtocolICMP = 1
+	// ProtocolIPv6ICMP is the ICMPv6 protocol number.
+	ProtocolIPv6ICMP = 58
+)
+
+// getICMPSeq returns the next sequence number as uint16,
+// wrapping around to 0 after reaching the maximum value of uint16.
+func getICMPSeq() uint32 {
+	// Increment the sequence number atomically and get the new value.
+	// We use atomic.AddUint32 and pass 1 as the increment.
+	// The returned value is the new value post-increment.
+	newVal := atomic.AddUint32(&icmpSeq, 1)
+
+	return newVal
+}
+
+// NodeLatencyMonitor is a tool to monitor the latency of the node.
+type NodeLatencyMonitor struct {
+	// Node config
+	nodeConfig *config.NodeConfig
+	// latencyStore is the cache to store the latency of each nodes.
+	latencyStore *LatencyStore
+	// latencyConfig is the config for the latency monitor.
+	latencyConfig *LatencyConfig
+	// latencyConfigChanged is the channel to notify the latency config changed.
+	latencyConfigChanged chan struct{}
+	// gatewayConfig is the traffic encap mode for the latency monitor,
+	// indicate if the Antrea Agent is running in network policy only mode.
+	trafficEncapMode config.TrafficEncapModeType
```

Comment:
same comment as above for `nodeConfig`: I don't think you need to store this in the struct